app/main.py

```diff
@@ -88,6 +91,29 @@ def _get_client(self):
                                              self.client_id, self.client_secret)
         return KeyVaultClient(credentials)
 
+    def _get_tenant_id(self):
+        vault_base_url = os.getenv('VAULT_BASE_URL')
+        _logger.info('Detecting tenant id for %s', vault_base_url)
+        # Some key to trigger a 401
+        URL = '{}/keys/somekeyname/1?api-version=2018-02-14'.format(vault_base_url)
+        _logger.info('Sending challenge request to %s', URL)
+        response = requests.get(url = URL)
+        if response.status_code == 401:
+            # If status code == HTTP 401, then parse the WWW-Authenticate header
+            # Bearer authorization="https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47", resource=".."
+            # Get the value of authorization and populate in self.tenant_id
+            # if conditions not met return the default tenant_id from cloud config file
+            challenge = response.headers['WWW-authenticate'].lower().replace('bearer ', '').split(',')
+            for kvp in challenge:
+                keyvalue = kvp.strip().split('=')
+                if len(keyvalue) == 2 and keyvalue[0] == 'authorization':
+                    authority = keyvalue[1].replace('"', '')
+                    tenant_id = urlparse(authority).path.replace('/', '')
+                    _logger.info('Tenant in config : %s, Tenant detected : %s. Using %s', self.tenant_id, tenant_id, tenant_id)
+                    self.tenant_id = tenant_id
+                    break
+        return self.tenant_id
```

Comment:
when this code becomes optional, and you enable it, you should decide what happens if you don't find the tenant_id in the bearer token.. should this fail or revert back to config? also, you should write it to log that you didn't find one

Reply:
yes. It falls back to the config default if it cannot detect the tenant id.